### PR TITLE
[List] Add spacing property

### PR DESCRIPTION
**Pull request by acidio**
### WHY are these changes introduced?

This PR is a follow up of[compact list](https://shopify.slack.com/archives/C4Y8N30KD/p1670936155108179) [the conversation about having a more](https://shopify.slack.com/archives/C4Y8N30KD/p1671128757572269?thread_ts=1670952079.108219&cid=C4Y8N30KD) for places in admin where we have limited space and where the page can get really long because of the amount of items on it.

<details>
      <summary>Current example</summary>

current

</details>

<details>
      <summary>Intended example</summary>

intended

</details>

### WHAT is this pull request doing?

Adding a `spacing` property to the `List` component allowing us to have a more compact version of the list.

The `spacing` options are `extraTight | loose`, the last been the default one so it doesn't break current usage.

| Before | After |
| - | - |
| image | image |

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';

import {List, Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      <List spacing="extraTight">
        <List.Item>Yellow shirt</List.Item>
        <List.Item>Red shirt</List.Item>
        <List.Item>Green shirt</List.Item>
      </List>
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/src/components/List/List.scss [hunk 1]
```diff
@@ -25,3 +23,9 @@
     margin-top: var(--p-space-2);
   }
 }
+
+.spacingLoose {
```
Comment (laurkim) [inline polaris-react/src/components/List/List.scss:23]: This probably needs to reset `margin-bottom` to `0` on the last child as well. It's what's causing the [unintended UI changes](https://shopify.chromatic.com/build?appId=5d559397bae39100201eedc1&number=14920) for related components with extra padding on the last list item.

**Review by laurkim — DISMISSED** [review]
Added a comment above about the margin for the last item in lists that needs to be updated to resolve the unintended UI changes it's causing.

File: polaris-react/src/components/List/List.scss (hunk 1 above)
Comment (acidio) [reply polaris-react/src/components/List/List.scss:23]: Oh, I forgot polaris has chromatics, thank you for point it out!

Comment (acidio) [reply polaris-react/src/components/List/List.scss:23]: After the last commit it's looking way better. b531b257c8cb8f0575015c19b79cf2193fa87f47

File: .changeset/weak-insects-jump.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@shopify/polaris': minor
+---
+
+Add spacing property to List component allowing for a more compact list
```
Comment (laurkim) [inline .changeset/weak-insects-jump.md]: Changesets are usually written in past tense

```suggestion
Added support for `spacing` prop to List component allowing for a more compact list
```

File: polaris-react/src/components/List/List.tsx [hunk 3]
```diff
@@ -7,7 +7,14 @@ import styles from './List.scss';
 
 type Type = 'bullet' | 'number';
 
+type Spacing = 'extraTight' | 'loose';
+
 export interface ListProps {
+  /**
+   * Determines the space between list items
+   * @default 'loose'
```
Comment (laurkim) [inline polaris-react/src/components/List/List.tsx:15]: Tysm for adding context with JSDoc! 💯 🌟

**Review by laurkim — APPROVED** [review]
Changes lgtm 👍 
I ran the `yarn get-props` script for our style guide to pick up the new spacing prop for the `List` component. There were merge conflicts on the `props.json` file when I tried to do it so I had to rebase this branch off latest `main`.

Updated `List` component page:
21-18-silrs-ehgl6

**Comment by acidio** [comment]
> Changes lgtm 👍 I ran the `yarn get-props` script for our style guide to pick up the new spacing prop for the `List` component. There were merge conflicts on the `props.json` file when I tried to do it so I had to rebase this branch off latest `main`.
> 
> Updated `List` component page: 21-18-silrs-ehgl6

Oh, thank you, I didn't know about this command, TIL :)

**Comment by acidio** [comment]
Closes https://github.com/Shopify/polaris/issues/5791

**Pull request merged**